scenario/mocking.py

```diff
@@ -688,19 +717,36 @@ def _layers(self) -> Dict[str, pebble.Layer]:
     def _service_status(self) -> Dict[str, pebble.ServiceStatus]:
         return self._container.service_status
 
-    def exec(self, *args, **kwargs):  # noqa: U100 type: ignore
-        cmd = tuple(args[0])
-        out = self._container.exec_mock.get(cmd)
+    # Based on a method of the same name from ops.testing.
+    def _find_exec_handler(self, command) -> Tuple[Optional[str], Optional["Exec"]]:
+        handlers = {
+            tuple(exc.command): (key, exc) for key, exc in self._container.execs.items()
+        }
+        for prefix_len in reversed(range(len(command) + 1)):
```

Comment:
not sure I understand what's going on. 
Suppose you have two execs, one for `ls -ll` and one for `ls -la`, and the charm does `ls --fubar`. Which exec mock will this match?

Reply:
The Harness behaviour (which I intended to match here, although I think I'm too low on tests) is that the longest match wins, but also that you have to match what's there. So in your example, neither would match.

| exec 1 | exec 2 | command | winner |
| - | - | - | - |
| ['ls', '-ll'] | ['ls', '-la'] | ['ls', '--fubar'] | none |
| [] | ['ls'] | ['ls',  '--fubar'] | exec 2|
| [] | ['ls', '-ll'] | ['ls', '--fubar'] | exec 1 |
| ['ls', '-ll', 'foo.txt'] | ['ls', '-ll'] | ['ls', '-ll', 'foo.txt'] | exec 1 |

I'm not sure what Harness does if there are two that have the exact same command pattern, but it seems to me that should be an error (consistency check?).

What I see charms doing (with Harness) is:

* using `[]`, presumably because they know that only one command will run and they either want to keep the test code clean or can't be bothered copying the exact command over to the test code
* using `[first]` when the command is things like `[first, arg1, arg2, arg2]` - probably partly the same reasoning as `[]` but maybe also because there are multiple ways that the charm could run it and the result should always be the same (I haven't looked deeply enough to check)
* using the full command

It does seem like having this is a nice convenience. @benhoyt probably has the full story on how Harness came to have that behaviour (it slightly predates me joining Canonical if I have the timing right).

Reply:
Talked over this with @benhoyt on a call also, and he agrees that it's worth keeping the prefix matching in. The [spec for adding exec testing in Harness](https://docs.google.com/document/d/1V7nzFzSOIMVlPF-HCvTZZyWOGlHnCS__RyL5352P2e4/edit) has some more of the background/details.